### ISSUE TITLE
[RFC] [DependencyInjection] Custom container configurators (fluent PHP format)

### ISSUE THREAD

**Issue by unkind** [symfony/symfony#25630]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | yes
| Symfony version  | 4.1 (?)

So far we have extensions for custom and simplified configurations, e.g.

```php
return function (ContainerConfigurator $c) {
    $c->extension('monolog', [
        'handlers' => [
            'main' => [
                'type'  => 'rotating_file',
                'path'  => '%kernel.logs_dir%/%kernel.environment%.log',
                'level' => 'debug',
                // max number of log files to keep
                // defaults to zero, which means infinite files
                'max_files' => 10,
            ],
        ],
    ]);
};
```

which is fine in many cases, but it doesn't use all capabilities of PHP. I'd be happy to have native custom configurator like the following, for example:

```php
return function (MonologContainerConfigurator $c) {
    $c->handlers()->register(
            'main',
            MonologHandlerType::rotatingFile()
                ->path('%kernel.logs_dir%/%kernel.environment%.log')
                ->maxFiles(10)
        )
        ->level(LoggerInterface::DEBUG);
};
```

I see at least 3 benefits:

1. It allows provide self-descriptive configurators, simplifies learning DSL of the extension (by playing with autocomplete options).
2. Such configurators prevent many typos because we work with methods, not custom arrays. So far only XML has self-validation, but this format is painful for many people.
3. It helps to introduce your own DSL in your project (e.g. I can add `->anonymous()` helper like I suggested in #24632). @nicolas-grekas mentioned it as possible solution for me:
    > It should be quite easy to define your own DSL, by creating a different set of configurator classes, and type-hinting for them in the closure.
   
     Unfortunately, Symfony doesn't support it out-of-box.

The idea is simple: bundle/extension should be able to register it's own container configurator factory:

```php
$container->registerFluentContainerConfigurator(
    MonologContainerConfigurator::class,
    function (ContainerConfigurator $c) {
        return new MonologContainerConfigurator($c);
    }
);
```

Then we can detect closure argument type and inject appropriate container configurator.

**Comment by iltar** [comment]
Though I really like the idea, won't this be double bookkeeping with the configuration nodes?

**Comment by mvrhov** [comment]
Maybe then we can deprecate the configuration

**Comment by nicolas-grekas** [comment]
I think you can already achieve what you want very easily:
```php
return function (ContainerConfigurator $c) {
    $myBundle = new MyBundleConfigurator($c);

    $myBundle->setWhatever();
}
```

Nothing required from core. WDYT?

**Comment by unkind** [comment]
> I think you can already achieve what you want very easily:

I think it's possible that bundle might want to inject some additional dependencies/parameters.
Also we have about hundred of config files in our project, if we want to use our custom configurator we have to copy-paste it many times, it's annoying.

**Comment by unkind** [comment]
> Though I really like the idea, won't this be double bookkeeping with the configuration nodes?

There is no violation of DRY if I understand you point correctly. Configuration nodes is attempt to make abstraction layer over all formats, but even in the current state it is leaky abstraction (e.g. there is `fixXmlConfig` method). It is OK for array-like configs, but fluent PHP is different. They may have similar keywords, nesting, etc., but it doesn't mean they are the same, so it's OK to support them separately. Strictly speaking, all formats (XML, YAML, fluent PHP, ...) are different [bounded] contexts and DRY makes sense within a single bounded context only.

**Comment by nicolas-grekas** [comment]
Fixed by #27065?

**Comment by unkind** [comment]
Oh, I see, technically it would work. It's just matter of taste I guess. Anonymous class with `extends` looks a little bit clumsy, but OK.

Still, I'd implement that feature (#25650) as well, but it's up to you.
